### PR TITLE
modified utp_weighted algo

### PR DESCRIPTION
**Pull request by dorotheapfeiffer**
### Issue reference / description

The branch you merge from should already reference an event-formation-unit github ticket number. You can add a descriptive title, but if an issue is referenced, you don't have to.

## Checklist for submitter

- [ ] Check for conflict with integration test
- [ ] Unit tests pass

---

## Nominate for Group Code Review (Anyone can nominate it)
Indicate if you think the code should be reviewed in a Thursday code review session.

- [ ] Recommend for group code review

Also, nominate it on the code_review Slack channel.

### CODE REVIEW

File: src/common/reduction/Cluster.cpp [hunk 1]
```diff
@@ -196,27 +181,46 @@ double Cluster::time_center2() const {
 double Cluster::coord_utpc(bool weighted) const {
   if (!weighted)
   {
-    return coord_utpc_/cnt_utpc_;
+    return coord_utpc_;
+  }
+
+  double n1 = 0, n2 = 0, n3 = 0, w1 = 0, w2 = 0, w3 = 0;
```
Comment (mortenjc) [inline src/common/reduction/Cluster.cpp]: You could make a case for using actual (and meaningful) names here :-)

File: src/common/reduction/Cluster.cpp [hunk 2]
```diff
@@ -196,27 +181,46 @@ double Cluster::time_center2() const {
...
+  if(highest_time_idx > 0) {
+    n1 = hits[highest_time_idx-1].coordinate;
+    w1 = hits[highest_time_idx - 1].weight;
+  } 
+  if(highest_time_idx < static_cast<int>(hits.size() - 1)) {
+    n3 = hits[highest_time_idx+1].coordinate;
+    w3 = hits[highest_time_idx+1].weight;  
   }
-  return coord_mass_utpc_ / weight_utpc_sum_;
+  double pos = (n1*w1*w1 + n2*w2*w2 + n3*w3*w3) / (w1*w1 + w2*w2 + w3*w3);
```
Comment (mortenjc) [inline src/common/reduction/Cluster.cpp]: but I see how this would affect this line. Nevertheless ....

File: src/common/reduction/Cluster.cpp [hunk 3]
```diff
@@ -196,27 +181,46 @@ double Cluster::time_center2() const {
...
+  return pos;
 }
 
 double Cluster::time_utpc(bool weighted) const {
   if (!weighted)
   {
-    return time_utpc_/cnt_utpc_;
```
Comment (mortenjc) [inline src/common/reduction/Cluster.cpp]: Was this a bug? (just curious)

File: src/common/reduction/analysis/test/EventAnalyzerTest.cpp [hunk 4]
```diff
@@ -85,45 +85,17 @@ TEST_F(EventAnalyzerTest, AnalyzeAverage) {
...
-
+  EXPECT_GT(result.center, 4.46);
+  EXPECT_LT(result.center, 4.47);
+  EXPECT_EQ(result.time, 4);
 }
 
-TEST_F(EventAnalyzerTest, AnalyzeUncert) {
-  hit.weight = 1;
-
-  hit.time = hit.coordinate = 0;
```
Comment (mortenjc) [inline src/common/reduction/analysis/test/EventAnalyzerTest.cpp]: I don't do this nearly enough myself, but consider adding some 
comments also to unit tests.

File: src/common/reduction/analysis/test/EventAnalyzerTest.cpp [hunk 5]
```diff
@@ -85,45 +85,17 @@ TEST_F(EventAnalyzerTest, AnalyzeAverage) {
...
-  EXPECT_EQ(result.center, 6);
+  EXPECT_EQ(result.center, 5);
   EXPECT_EQ(result.time, 5);
 
   result = EventAnalyzer("utpc_weighted").analyze(cluster);
-  EXPECT_EQ(result.center, 5.5);
-  EXPECT_EQ(result.time, 5);
-  
-
+  EXPECT_GT(result.center, 4.46);
```
Comment (mortenjc) [inline src/common/reduction/analysis/test/EventAnalyzerTest.cpp]: Maybe better to use
ASSERT_NEAR(val1, val2, abs_error);
?

File: src/common/reduction/analysis/test/EventAnalyzerTest.cpp (hunk 4 above)
Comment (mortenjc) [reply src/common/reduction/analysis/test/EventAnalyzerTest.cpp]: Of course not in deleted code, but you know what I mean :-)

File: src/common/reduction/analysis/test/EventAnalyzerTest.cpp (hunk 5 above)
Comment (mortenjc) [reply src/common/reduction/analysis/test/EventAnalyzerTest.cpp]: for floatsand doubles

File: src/common/reduction/Cluster.cpp (hunk 1 above)
Comment (dorotheapfeiffer) [reply src/common/reduction/Cluster.cpp]: The intention was: n = numerator, w = weight
The numerator is either the strip or the time, and the weight is the charge or adc  value.

File: src/common/reduction/Cluster.cpp (hunk 3 above)
Comment (dorotheapfeiffer) [reply src/common/reduction/Cluster.cpp]: In Martin's uTPC for APV algorithm, the latest times were used. If there were several strips with the same latest time, the average of the strip position was calculated. If the option was weighted uTPC, not the average of the strip positions was used, but the weighted average (using the charge). In my new implementation, pure uTPC just takes the strip with the largest time (with ns resolution there are basically never two strips with the same time). The weighted variant takes the strip with the latest time, together with it's one or two neighbours, and calculates center of mass with the charge squared as weight,

**Pull request merged**